src/net_processing.cpp

```diff
@@ -2039,7 +2076,16 @@ void PeerManagerImpl::NewPoWValidBlock(const CBlockIndex *pindex, const std::sha
 void PeerManagerImpl::UpdatedBlockTip(const CBlockIndex *pindexNew, const CBlockIndex *pindexFork, bool fInitialDownload)
 {
     SetBestHeight(pindexNew->nHeight);
-    SetServiceFlagsIBDCache(!fInitialDownload);
+
+    // Update internal sync state.
+    // Avoid using 'fInitialDownload' alone since the flag currently cannot go backwards after passing IBD once.
+    // This means that, after completing sync for the first time, 'fInitialDownload' will remain true even
+    // when the node has been stalled for days and the arriving block is within the IBD window.
+    if (!fInitialDownload && IsBlockCloseToTip(pindexNew, m_chainparams.GetConsensus())) {
+        // At this point, only move the flag forward. The responsibility of setting the flag backwards is on
+        // the 'CheckForStaleTip' process, which performs a more exhaustive stale tip verification.
```

Comment:
> and C should be just "our highest block is older than 48h" (regardless of whether mapBlocksInFlight is empty) => then don't connect to limited peers.

Hmm, this observation makes me think..
what if `m_is_close_to_tip=true` and then the node lags behind; there could be an edge case scenario where the node starts slowly syncing-up the historical blocks before `CheckForStaleTipAndEvictPeers()` is executed, so `mapBlocksInFlight` wouldn't be empty, so.. it would still connect to limited peers even when it is behind what they can provide.
This is what you have in mind?

Reply:
I was just thinking on ways to simplify this. `CheckForStaleTipAndEvictPeers()` is executed every 45 seconds. I think it is ok if for a while (~ 10 mins) we think we are close to the tip but we are not in practice. Given that we realize this in a few minutes and don't get stuck with the wrong assumption forever.